sys/kern/imgact_elf.c

```diff
@@ -999,6 +998,103 @@ __elfN(enforce_limits)(struct image_params *imgp, const Elf_Ehdr *hdr,
 	return (0);
 }
 
+
+static int
+__elfN(ape_find_printf)(const char *page, int i)
+{
+	for (; i + 8 < 4096; ++i) {
+		if (memcmp(page + i, "printf '", 8) == 0)
+			return (i + 8);
+	}
+	return (-1);
+}
+
+static int
+__elfN(ape_parse_octal)(const char *page, int i, int *pc)
+{
+	int c;
+	if ('0' <= page[i] && page[i] <= '7') {
+		c = page[i++] - '0';
+		if ('0' <= page[i] && page[i] <= '7') {
+			c *= 8;
+			c += page[i++] - '0';
+			if ('0' <= page[i] && page[i] <= '7') {
+				c *= 8;
+				c += page[i++] - '0';
+			}
+		}
+		*pc = c;
+	}
+	return (i);
+}
+
+/*
+ * Files beginning with "MZqFpD" are Actually Portable Executables,
+ * which have a printf statement in the first 4096 bytes with octal
+ * codes that specify the ELF header. APE also specifies `jartsr='`
+ * as an alternative prefix, intended for binaries that do not want
+ * to be identified as Windows executables. Like the \177ELF magic,
+ * all these prefixes decode as x86 jump instructions that could be
+ * used for 16-bit bootloaders or 32-bit / 64-bit flat executables.
+ * Most importantly they provide a fallback path for Thompson shell
+ * compatible command interpreters, which do not require a shebang,
+ * e.g. bash, zsh, fish, bourne, almquist, etc. Please note that in
+ * order to meet the requirements of POSIX.1, the single quote must
+ * be followed by a newline character, before any null bytes occur.
+ * See also: https://www.austingroupbugs.net/view.php?id=1250
+ */
+static int
+__elfN(ape_decode_elf)(const char *page, char *ebuf)
+{
+	int c;
+	int i;
+	int retval;
+	Elf_Ehdr *ehdr;
+	int ebuf_index;
+	int desired_machine;
+
+#if defined(__aarch64__)
+	desired_machine = EM_AARCH64;
+#elif defined(__powerpc64__)
+	desired_machine = EM_PPC64;
+#elif defined(__riscv)
+	desired_machine = EM_RISCV;
+#else
+	desired_machine = EM_X86_64;
+#endif
+
+	i = 0;
+keep_looking:
```

Comment:
Not a requirement, but this feels like it would be a pretty natural use of `do {...} while (ehdr->e_machine != desired_machine);`